controller/sharding/sharding.go

```diff
@@ -23,31 +33,83 @@ func InferShard() (int, error) {
 	if err != nil {
 		return 0, fmt.Errorf("hostname should ends with shard number separated by '-' but got: %s", hostname)
 	}
-	return shard, nil
+	return int(shard), nil
 }
 
-// GetShardByID calculates cluster shard as `clusterSecret.UID % replicas count`
-func GetShardByID(id string, replicas int) int {
-	if id == "" {
-		return 0
-	} else {
-		h := fnv.New32a()
-		_, _ = h.Write([]byte(id))
-		return int(h.Sum32() % uint32(replicas))
+func GetClusterFilter(distributionFunction DistributionFunction, shard int) func(c *v1alpha1.Cluster) bool {
+	return func(c *v1alpha1.Cluster) bool {
+		return distributionFunction(c) == shard
 	}
 }
 
-func GetClusterFilter(replicas int, shard int) func(c *v1alpha1.Cluster) bool {
-	return func(c *v1alpha1.Cluster) bool {
-		clusterShard := 0
-		//  cluster might be nil if app is using invalid cluster URL, assume shard 0 in this case.
+func GetDistributionFunction(kubernetesClient *kubernetes.Clientset, settingsMgr *settings.SettingsManager) func(c *v1alpha1.Cluster) int {
+	filterFunctionName := env.StringFromEnv(common.EnvControllerShardingAlgorithm, "legacy")
+	distributionFunction := GetShardByIdUsingHashDistributionFunction()
```

Comment:
let's store this as a variable

Reply:
Which one @ishitasequeira ?
these are 2 variables that you are highlighting.